src/main/java/org/javarosa/core/model/FormDef.java

```diff
@@ -166,7 +176,8 @@ public static IDataReference getAbsRef(IDataReference ref, TreeReference parentR
 
     private HashMap<String, SubmissionProfile> submissionProfiles;
 
-    private HashMap<String, DataInstance> formInstances;
+    private HashMap<String, DataInstance> externalFormInstances = new HashMap<>();
```

Comment:
Would you consider keeping a single map here? I believe the serialization methods are the only ones that need to know the difference between internal and external instances so a single map seems it would better match that narrow usage. 

In that case, the only method that would need to know about the difference between internal and external form instances is `getExternalInstances`. It would be the one to filter out only the external ones. `addNonMainInstance`, the constructor, etc, could then all stay the same and you wouldn't need `getInternalInstances`.